mempool/iterators.go

```diff
@@ -7,37 +7,44 @@ import (
 	"github.com/cometbft/cometbft/types"
 )
 
-// WRRIterator is the base struct for implementing iterators that traverse lanes with
-// the classical Weighted Round Robin (WRR) algorithm.
-type WRRIterator struct {
-	sortedLanes []types.Lane
-	laneIndex   int                            // current lane being iterated; index on sortedLanes
-	counters    map[types.Lane]uint            // counters of consumed entries, for WRR algorithm
-	cursors     map[types.Lane]*clist.CElement // last accessed entries on each lane
+// IWRRIterator is the base struct for implementing iterators that traverse lanes with
+// the Interleaved Weighted Round Robin (WRR) algorithm.
+// https://en.wikipedia.org/wiki/Weighted_round_robin
+type IWRRIterator struct {
+	sortedLanes  []types.Lane
+	laneIndex    int                            // current lane being iterated; index on sortedLanes
+	cursors      map[types.Lane]*clist.CElement // last accessed entries on each lane
+	roundCounter int                            // counts the rounds for IWRR
 }
 
-func (iter *WRRIterator) nextLane() types.Lane {
+func (iter *IWRRIterator) nextLane() types.Lane {
+	if iter.laneIndex == len(iter.sortedLanes)-1 {
+		iter.roundCounter = (iter.roundCounter + 1) % (int(iter.sortedLanes[0]) + 1)
+		if iter.roundCounter == 0 {
+			iter.roundCounter++
+		}
+	}
```

Comment:
I tied this to the laneIndex increment because the round increment happens when all lanes have been processed. So we have to have a way to check whether we have gone through all the lanes before increasing the round number. That's why I chose to do it all in one place.